### PR TITLE
[Curp] Log entry persistency

### PR DESCRIPTION
**Pull request by markcty**
Based on #178 

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Persist log entries and `voted_for`.

* what changes does this pull request make?
This PR also rename `ServerTimeout` config to `CurpConfig` since now it will contain data dir.

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: curp/src/server/raw_curp/mod.rs [hunk 1]
```diff
@@ -612,6 +613,58 @@ impl<C: 'static + Command> RawCurp<C> {
...
+        uncommitted_pool: UncommittedPoolRef<C>,
+        cfg: Arc<CurpConfig>,
+        cmd_tx: Box<dyn CmdExeSenderApi<C>>,
+        sync_tx: mpsc::UnboundedSender<usize>,
+        calibrate_tx: mpsc::UnboundedSender<ServerId>,
+        log_tx: mpsc::UnboundedSender<LogEntry<C>>,
+        voted_for: Option<(u64, ServerId)>,
+        entries: Vec<LogEntry<C>>,
+    ) -> Self {
+        // recovered cmds should be stored in the
```
Comment (Phoenix500526) [inline curp/src/server/raw_curp/mod.rs:634]: In what?

File: curp/src/server/raw_curp/mod.rs [hunk 2]
```diff
@@ -612,6 +613,58 @@ impl<C: 'static + Command> RawCurp<C> {
...
+            calibrate_tx,
+            log_tx.clone(),
+        );
+
+        if let Some((term, server_id)) = voted_for {
+            let mut st_w = raw_curp.st.write();
+            raw_curp.update_to_term_and_become_follower(&mut st_w, term);
+            st_w.voted_for = Some(server_id);
+        } else if is_leader {
+            // all uncommitted cmds should stay in ucp until there are executed
```
Comment (Phoenix500526) [inline curp/src/server/raw_curp/mod.rs:654]: Typo? 'until they are executed'

File: curp/src/server/raw_curp/log.rs [hunk 3]
```diff
@@ -22,17 +24,34 @@ pub(super) struct Log<C: Command> {
...
     /// Create a new log
-    pub(super) fn new() -> Self {
+    pub(super) fn new(
+        log_tx: mpsc::UnboundedSender<LogEntry<C>>,
+        entries: Vec<LogEntry<C>>,
+    ) -> Self {
         Self {
-            entries: vec![], // a fake log[0]
+            entries, // a fake log[0]
             commit_index: 0,
```
Comment (themanforfree) [inline curp/src/server/raw_curp/log.rs:50]: why we didn't recover those indexes?

File: curp/src/server/raw_curp/mod.rs (hunk 1 above)
Comment (markcty) [reply curp/src/server/raw_curp/mod.rs:634]: Fixed.

File: curp/src/server/raw_curp/mod.rs [hunk 4]
```diff
@@ -571,10 +571,11 @@ impl<C: 'static + Command> RawCurp<C> {
         cmb_board: CmdBoardRef<C>,
```
[the remark this answers is not in the document]
Comment (markcty) [reply curp/src/server/raw_curp/mod.rs:571]: fixed

File: curp/src/server/raw_curp/log.rs (hunk 3 above)
Comment (markcty) [reply curp/src/server/raw_curp/log.rs:53]: Last applied should be maintained by command executor while commit index can be fetched from the leader as soon as it recovers

**Comment by markcty** [comment]
Well, I don't think the coverage can be improved in this PR.

**Review by Phoenix500526 — APPROVED** [review]

**Pull request merged**